File: disjoint-set/python/disjoint_set_with_list.py
class Element:
    __slots__ = ["data", "next", "prev"]

    def __init__(self, data):
        self.data = data
        self.next = self.prev = None


class LinkedList:
    __slots__ = ["head", "tail"]

    def __init__(self):
        self.head = self.tail = None


def make_set(data):
    ll = LinkedList()
    ll.head = ll.tail = Element(data)

    return ll


def union(ch1, ch2, sets):
    parent_of_x = find_set(ch1, sets)
    parent_of_y = find_set(ch2, sets)

    if parent_of_x == parent_of_y:
        return

    x = sets[parent_of_x]
    y = sets[parent_of_y]

    if parent_of_x < parent_of_y:
        x.tail.next = y.head
        x.tail = y.tail
        del sets[ch2]


def find_set(data, sets):
    for v in sets.values():
        found = False
        p = v.head
        while p:
            if p.data == data:
                found = True

                break

            p = p.next

        if found:
            return v.head.data

    return None


def print_sets(sets):
    for k, v in sets.items():
        print(k, ":", end=" ")
        p = v.head
        while p:
            print(p.data, end=" -> " if p.next else "")

            p = p.next

        print("")


if __name__ == "__main__":
    characters = "abcdefghij"

    sets = {}
    for ch in characters:
        sets[ch] = make_set(ch)

    print("--- after initialize ---")
    print_sets(sets)

    print("union b and d")
    union("b", "d", sets)
    print_sets(sets)
    print("union e and g")
    union("e", "g", sets)
    print_sets(sets)
    print("union a and c")
    union("a", "c", sets)
    print_sets(sets)
    print("union h and i")
    union("h", "i", sets)
    print_sets(sets)
    print("union a and b")
    union("a", "b", sets)
    print_sets(sets)
    print("union e and f")
    union("e", "f", sets)
    print_sets(sets)
    print("union b and c")
    union("b", "c", sets)
    print_sets(sets)

File: disjoint-set/golang/disjoint_set_with_list.go
package main

import "fmt"

type Element struct {
	data       string
	next, prev *Element
}

type LinkedList struct {
	head, tail *Element
}

func makeSet(data string) *LinkedList {
	ll := &LinkedList{}

	e := &Element{data: data}
	ll.head = e
	ll.tail = e

	return ll
}

func printSets(sets map[string]*LinkedList) {
	for k, v := range sets {
		fmt.Printf("%v : ", k)

		for p := v.head; p != nil; p = p.next {
			fmt.Printf("%v", p.data)
			if p.next != nil {
				fmt.Print(" -> ")
			}
		}

		fmt.Print("\n")
	}
}

func findSet(ch string, sets map[string]*LinkedList) string {
	for _, v := range sets {
		found := false
		for p := v.head; p != nil; p = p.next {
			if p.data == ch {
				found = true

				break
			}
		}

		if found {
			return v.head.data
		}
	}

	return ""
}

func union(ch1, ch2 string, sets map[string]*LinkedList) {
	parentOfX := findSet(ch1, sets)
	parentOfY := findSet(ch2, sets)

	if parentOfX == parentOfY {
		return
	}

	x := sets[parentOfX]
	y := sets[parentOfY]

	if parentOfX < parentOfY {
		x.tail.next = y.head
		x.tail = y.tail
		delete(sets, parentOfY)
	}
}

func main() {
	sets := make(map[string]*LinkedList)

	for _, ch := range "abcdefghi" {
		sets[string(ch)] = makeSet(string(ch))
	}

	fmt.Println("--- initialize ---")
	printSets(sets)

	fmt.Println("union b and d")
	union("b", "d", sets)
	printSets(sets)
	fmt.Println("union e and g")
	union("e", "g", sets)
	printSets(sets)
	fmt.Println("union a and c")
	union("a", "c", sets)
	printSets(sets)
	fmt.Println("union h and i")
	union("h", "i", sets)
	printSets(sets)
	fmt.Println("union a and b")
	union("a", "b", sets)
	printSets(sets)
	fmt.Println("union e and f")
	union("e", "f", sets)
	printSets(sets)
	fmt.Println("union b and c")
	union("b", "c", sets)
	printSets(sets)
}
